File: js/collatz.js
var callstack;
var caching;
var cache;

function collatz(value, steps) {
  benchmark();
  if (cache[value] && caching){
    return cache[value] + steps - 1;
  } else if (value === 1){
    return steps;
  } else if (value % 2){
    return collatz(3 * value + 1, steps + 1);
  } else {
    return collatz(value / 2, steps + 1);
  }
}

function calculate_collatz_steps(integer, c) {
  callstack = 0;
  caching = c;
  cache = {};

  for (i = 1; i <= integer; i++){
    cache[i] = collatz(i, 1);
  }
  var result = Math.max.apply(null,
    Object.keys(cache).map(function(e) {
      return cache[e];
    }));
  console.log(
    "Iterations: " + callstack +
    " integer: " + cache[result] +
    " steps: " + result
  );
}

function benchmark() {
  callstack += 1;
}

integer = 10000;

calculate_collatz_steps(integer, false);
calculate_collatz_steps(integer, true);

File: ruby/collatz.rb
def collatz(value,steps)
  benchmark
  if @cache[value] && @caching
    return @cache[value] + steps - 1
 elsif value == 1
    return steps
 elsif value % 2 == 0
    return collatz(value/2,steps+1)
  else
    return collatz(3*value+1,steps+1)
  end
end

def calculate_collatz_steps(range, caching)
  @callstack = 0
  @caching = caching
  @cache = Hash.new
  range.each do |value|
    @cache[value] = collatz(value,1)
  end
  puts "Iterations: #{@callstack} integer: #{@cache.invert[@cache.values.max]} steps:  #{@cache[@cache.invert[@cache.values.max]]}"
end

def benchmark
  @callstack+=1
end

range = (1..10000)

calculate_collatz_steps(range, false)
calculate_collatz_steps(range, true)